cli/remote_build.go

```diff
@@ -886,45 +903,89 @@ func (cmd *RemoteCmd) submitBuild(project *Project, tagMap map[string]string) er
 	log.Infof("Submission finished at %s, taking %s", endTime.Format(TIME_FORMAT), submitTime.Truncate(time.Millisecond))
 
 	startTime = time.Now()
-	var remoteProgress *Progress
-	remoteErrored := func() {
-		if remoteProgress != nil {
-			remoteProgress.Fail()
-		}
-	}
 	if log.CheckIfTerminal() {
 		remoteProgress = NewProgressSpinner("Setting up remote build")
 		remoteProgress.Start()
 	}
 
 	if strings.HasPrefix(url, "ws:") || strings.HasPrefix(url, "wss:") {
+		recentReconnect := false
+		reconnectCount := 0
+	CONN:
 		conn, _, _, err := ws.DefaultDialer.Dial(context.Background(), url)
+
+		finnish := make(chan struct{})
+
 		if err != nil {
 			remoteErrored()
 			return fmt.Errorf("Cannot connect: %v", err)
 		} else {
 
+			// TODO maybe make Dispatcher give better diagnostics somehow
+			go func() {
+				for {
+					select {
+					case <-finnish:
+						return
+					case <-time.After(5 * time.Second):
+						if err := wsutil.WriteClientMessage(conn, ws.OpPing, []byte("hi Dispatcher")); err != nil {
+							log.Errorf("Cannot send ping: %v", err)
+						}
+					}
+				}
+			}()
+
 			defer func() {
 				if err = conn.Close(); err != nil {
 					log.Debugf("Cannot close: %v", err)
 				}
 			}()
 
 			buildSuccess := false
+			buildFailed := false
 			buildSetupFinished := false
 			for {
 				msg, control, err := wsutil.ReadServerData(conn)
 				if err != nil {
-					if err != io.EOF {
-						log.Tracef("Unstable connection: %v", err)
-					} else {
+					if err == io.EOF {
 						if buildSuccess {
 							log.Infoln("Build Completed!")
+							close(finnish)
+							return nil
+						} else if buildFailed {
+							log.Errorf("Build failed!")
+							log.Infof("Build Log: %v", managementLogUrl(url, project.OrgSlug, project.Label))
+
+							close(finnish)
+							return nil
 						} else {
-							log.Errorln("Build failed or the connection was interrupted!")
+							if !recentReconnect && reconnectCount < 15 {
+								if remoteProgress != nil {
+									fmt.Println()
+								}
+								log.Tracef("Build not completed, trying to reconnect")
+								conn.Close()
+								close(finnish)
+								recentReconnect = true
+								reconnectCount += 1
+								goto CONN
```

Comment:
goto can be a little dangerous / hard to grok; any way to avoid it here? I can't tell for sure if we can - if it's not easy then this is fine at the moment

Reply:
I don't like it too. Maybe I'll rewrite the whole remotebuild stuff after runtime-context is merged, because it got a little too long. I was thinking of trying to replace `subcommands` with something like `cobra` to manage the CLI subcommands, log output and maybe follow @amyhua's idea on a better CLI tool.